src/libraries/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.CurrentUserOnly.Windows.cs

```diff
@@ -131,14 +131,51 @@ public NamedPipeTest_CurrentUserOnly_Windows(TestAccountImpersonator testAccount
             _testAccountImpersonator = testAccountImpersonator;
         }
 
+        [OuterLoop]
+        [ConditionalFact(nameof(IsAdminOnSupportedWindowsVersions))]
+        public async Task Connection_UnderDifferentUsers_CurrentUserOnlyOnServer_DoesNotBlockOthers()
+        {
+            string name = PipeStreamConformanceTests.GetUniquePipeName();
+            bool invalidClientShouldStop = false;
+
+            using var server = new NamedPipeServerStream(name, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly);
+            Task serverWait = server.WaitForConnectionAsync();
+
+            Task invalidClient = Task.Run(() =>
+            {
+                // invalid non-current user tries to connect to server.
+                _testAccountImpersonator.RunImpersonated(() =>
+                {
+                    while (!Volatile.Read(ref invalidClientShouldStop))
+                    {
+                        using var client = new NamedPipeClientStream(".", name, PipeDirection.In, PipeOptions.Asynchronous);
+                        Assert.Throws<UnauthorizedAccessException>(() => client.Connect());
+                    }
+                });
+            });
+
+            Thread.Sleep(1000); // give it a sec to allow above tasks to perform some work before this.
```

Comment:
I've gotten pretty suspicious of sleeps in tests! I guess it's OK since the test won't fail on a slow machine, but the ideal pattern is to wait on some event set in the task eg after it's looped.

Reply:
in fact, rather than having one task looping, you could consider kicking off N tasks all trying to connect once. That will give you some concurrency, and you can just do Task.WaitAll for them to complete, no sleep and no event.

Reply:
the WaitAll would be after the `await serverWait;`. 

As an improvement, I suggest to have the tasks all wait on an event before they try to connect, and then set that event on the line before you do `client.Connect()`. That will make it most likely they do their work concurrently with the succesful connect.

Reply:
+1

Reply:
With `serverWait = server.WaitForConnectionAsync()`, we can't do anything _after_ awaiting it, the method blocks until a client connects; the event should be set _before_ awaiting serverWait. Please check latest commit.

Reply:
> I've gotten pretty suspicious of sleeps in tests!

Sometimes they're necessary.  They're erroneous when the sleep is used to guarantee that something has or hasn't happened and the test will spuriously fail if the timeout ends up being too short, but that's not the case here.  This is a case where the sleep is trying to allow something to happen, 99.999% of the time the sleep will successfully delay things to the point where it does happen, and in the remaining 0.001% of the time that it doesn't, the test still passes, just not exercising the desired interleaving.

Reply:
> With serverWait = server.WaitForConnectionAsync(), we can't do anything after awaiting it, the method blocks until a client connects; the event should be set before awaiting serverWait. Please check latest commit.

The concern here is essentially whenever you spawn an asynchronous task, as `Task.Run()` is doing, you cannot guess at how long it will take to start running code. Which is why you need to set an event inside the body, eg., just before `Assert.Throws<UnauthorizedAccessException>(() => client.Connect());` ? The event means "the loop is executing!". Then just before the `client.Connect()` below that we expect to succeed, wait on that event.

Reply:
@stephentoub would that address your feedback, then this could be merged?

Reply:
Even setting an event just before it's invoked has a race condition.  This is the difficulty with synchronous methods; there's no way to signal "I'm now executing" because by the time you are you can't do anything else.  Hence why I suggested a loop to increase the chances of collision.

But, if you're really against the loop, what you suggest is the next best thing.

Reply:
I'm not against the loop, I believe I suggested it in another comment. Over to you @Jozkee 🙂

Reply:
Another way to circumvent the Thread.Sleep could be to wait on the loop say 100 iterations, then proceed to connect with the valid client. That way we ensure that the invalid client is running and there won't be a fixed time lock.

Reply:
Not sure I follow. How is 100 better than 1? The ideal is to not have any guesses.

Reply:
I'n not very good explaining so, here's how the test would look like:
```diff
[OuterLoop]
[ConditionalFact(nameof(IsAdminOnSupportedWindowsVersions))]
public async Task Connection_UnderDifferentUsers_CurrentUserOnlyOnServer_InvalidClientConnectionAttempts_DoNotBlockSuccessfulClient()
{
    string name = PipeStreamConformanceTests.GetUniquePipeName();
    bool invalidClientShouldStop = false;
+    int invalidClientAttempts = 0;

    using var server = new NamedPipeServerStream(name, PipeDirection.InOut, 1, PipeTransmissionMode.Byte, PipeOptions.Asynchronous | PipeOptions.CurrentUserOnly);
    Task serverWait = server.WaitForConnectionAsync();

    Task invalidClient = Task.Run(() =>
    {
        // invalid non-current user tries to connect to server.
        _testAccountImpersonator.RunImpersonated(() =>
        {
            while (!Volatile.Read(ref invalidClientShouldStop))
            {
                using var client = new NamedPipeClientStream(".", name, PipeDirection.In, PipeOptions.Asynchronous);
                Assert.Throws<UnauthorizedAccessException>(() => client.Connect());
+                Volatile.Write(ref invalidClientAttempts, invalidClientAttempts + 1);
            }
        });
    });

-    Thread.Sleep(1000); // give it a sec to allow above tasks to perform some work before this.
+    while (Volatile.Read(ref invalidClientAttempts) < 100) ; // Wait until the invalid client has tried multiple times.

    Assert.False(serverWait.IsCompleted);

    // valid client tries to connect and succeeds.
    using var client = new NamedPipeClientStream(".", name, PipeDirection.In, PipeOptions.Asynchronous);
    client.Connect();
    await serverWait;
    Volatile.Write(ref invalidClientShouldStop, true);
}
```

Reply:
What I am arguing is that we can have the loop with the event, and without the thread.sleep.

Reply:
Yes to the loop without the sleep, but what I'm asking is why you need a counter. As long as it's gone through the loop once, you know the "interference" is in progress, right? We don't expect any new behavior to emerge after 100 times.

Reply:
Yes, I agree there's no need for 100 times.